### PR TITLE
feat: Unique filters plugin

### PR DESCRIPTION
**Pull request by LoveIsGrief**
Check-list
----------

 * [x] There is a linked issue discussing the motivations for this feature or bugfix --> https://github.com/quodlibet/quodlibet/discussions/4222
 * [x] Unit tests have been added where possible 
 > There don't seem to be any tests for plugins?
 * [x] I've added / updated documentation for any user-facing features 
 > Is the plugin description enough?
 * [x] Performance seems to be comparable or better than current `main`

What this change is adding / fixing
-----------------------------------

This plugin allows filtering by a unique tag e.g `@(unique: artist)` which will result only one song showing up per artist. It can be used multiple times e.g if one has mix albums in the library (albums with multiple artists),
 it's possible to use `&(@(unique: artist), @(unique: album))` to have one song per artist and one song per album

### CODE REVIEW

**Comment by LoveIsGrief** [comment]
This should be ready now :slightly_smiling_face:

File: quodlibet/ext/query/unique.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+            return_value = field_value not in self.unique_tag_values
+            self.unique_tag_values.add(field_value)
+            print_d(f"unique filtering value '{field_value}': {return_value}")
+        except KeyError:
+            pass
+        except Exception as e:
+            err_str = str(e)
+            if err_str not in self._reported:
+                self._reported.add(err_str)
+                print_w("%s while filtering unique values for '%s': %s "
```
Comment (declension) [inline quodlibet/ext/query/unique.py:43]: Generally we're moving to f-strings for new code

File: quodlibet/ext/query/unique.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    PLUGIN_DESC = _("Filter search results by unique tags.")
+    key = 'unique'
+    query_syntax = _("@(unique: tag)")
+    query_description = "<tt>tag</tt> can be album, artist, title or any other tag. " \
+        "Use multiple <tt>@(unique: tag)</tt> to filter by multiple tags."
+
+    usage = f"{query_syntax}\n\n{query_description}"
+
+    def __init__(self):
+        print_d("Initialising unique query")
```
Comment (declension) [inline quodlibet/ext/query/unique.py:24]:
```suggestion
```
Probably not necessary any more

File: quodlibet/ext/query/unique.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+        self.unique_tag_values = set()
+        """The unique tag values that have been seen in the songs being filtered"""
+        self._reported = set()
+        """Unique errors to counter error log spam"""
+
+    def search(self, data, body):
```
Comment (declension) [inline quodlibet/ext/query/unique.py:30]: Could you add some type annotations? Trying to do that for all / nearly all new code

**Review by declension — APPROVED** [review]
LGTM generally, thanks!

File: quodlibet/ext/query/unique.py (hunk 3 above)
Comment (LoveIsGrief) [reply quodlibet/ext/query/unique.py]: I added it for the method signature. Dunno if variables and such also need them. If would be to have `mypy` check them and throw errors for where you want type hints to be put.

File: quodlibet/ext/query/unique.py (hunk 1 above)
Comment (LoveIsGrief) [reply quodlibet/ext/query/unique.py]: Done

File: quodlibet/ext/query/unique.py (hunk 2 above)
Comment (LoveIsGrief) [reply quodlibet/ext/query/unique.py]: Removed

**Comment by LoveIsGrief** [comment]
Thanks for the review @declension . I hope it's alright now.

**Comment by declension** [comment]
Thanks!

**Pull request merged**